### DISCUSSION TITLE
terragrunt references the wrong hashicorp/aws version

### DISCUSSION THREAD

**Discussion by mfractal** [gruntwork-io/knowledge-base#692] [Help]
Hi all, some background

we have a root terragrunt.hcl file with the following, notice the lack of version declaration

```
# Generate an AWS provider block
generate "provider" {
  path      = "provider.tf"
  if_exists = "overwrite_terragrunt"
  contents  = <<EOF
provider "aws" {
  region = "${local.aws_region}"
}
EOF
}
```

Each module then runs 

```
include "root" {
  path = find_in_parent_folders()
}
```

The problem :

for some reason the storage module, insists on importing a lower version of aws than what we need and i get tons of errors on some s3 declarations.

storage/terragrunt.lock.hcl :
```
provider "registry.terraform.io/hashicorp/aws" {
  version     = "3.74.3"
  constraints = "~> 3.74.1"
```

While for example in kubernetes module :
```
provider "registry.terraform.io/hashicorp/aws" {
  version     = "4.47.0"
  constraints = ">= 3.72.0, >= 4.0.0"
```

I tried deleting all .terragrunt-cache folders, tried running init and init -upgrade but it keeps insisting on 3.74.3 for storage and i can't for the life of me figure out why. 

P.S
initially i was getting the following error 
```
Initializing the backend...

Initializing provider plugins...
- Reusing previous version of hashicorp/aws from the dependency lock file
╷
│ Error: Failed to query available provider packages
│ 
│ Could not retrieve the list of available versions for provider
│ hashicorp/aws: locked provider registry.terraform.io/hashicorp/aws 4.51.0
│ does not match configured version constraint ~> 3.74.1; must use terraform
│ init -upgrade to allow selection of new versions

```

But then i ran 
```
terragrunt init -upgrade     

Initializing the backend...

Initializing provider plugins...
- Finding hashicorp/aws versions matching "~> 3.74.1"...
- Installing hashicorp/aws v3.74.3...
- Installed hashicorp/aws v3.74.3 (signed by HashiCorp)
```

and as you can see it pulled 3.74 instead of 4. 
Please advise :) 

---

<ins datetime="2023-03-31T15:39:10Z">
  <p><a href="https://support.gruntwork.io/hc/requests/110043">Tracked in ticket #110043</a></p>
</ins>

**Comment by denis256** [comment]
Hi,
will be helpful if will be mentioned used providers versions or better provided an example repo where the issue occurs

**Reply by mfractal** [reply]
hi! what do you mean used providers versions ? i thought i mentioned it above.  i am happy to provide more info, just not sure what is asked of me here :)

**Reply by denis256** [reply]
Hello,
from provided details is not clear which 3rd party modules are used which may have own `required_providers` redefined

Can be searched for `required_providers` in `.terragrunt-cache` to see if other files re-define `required_providers` version

To override versions can be attempted to generate `required_providers` from the parent `terragrunt.hcl` (or to generate empty override file to disable setting from module):
```

generate "versions" {
  path      = "versions_override.tf"
  if_exists = "overwrite_terragrunt"
  contents  = <<EOF
    terraform {
      required_providers {
        aws = {
          source = "hashicorp/aws"
          version = "4.60.0"
        }
      }
    }
EOF
}

```

https://developer.hashicorp.com/terraform/language/providers/requirements#requiring-providers

https://github.com/denis256/terragrunt-tests/blob/master/discussion-692/terragrunt.hcl
